### PR TITLE
flesh out controller ux

### PR DESCRIPTION
**Pull request by willscott**
use esbuild over an npm package for frontend js

image

### CODE REVIEW

File: Dockerfile [hunk 1]
```diff
@@ -6,8 +12,10 @@ RUN apk add --update gcc>=9.3.0 g++>=9.3.0 alpine-sdk
 WORKDIR /go/src/app/
 
 COPY . .
+COPY --from=js /usr/src/app ./controller/app
 # Fetch dependencies.
 RUN go get -d -v ./...
+RUN go generate ./...
```
Comment (mvdan) [inline Dockerfile:18]: should this not be committed to master as-is? if we're worrying about forgetting to re-run go generate, maybe we should make that a CI check instead.

File: controller/app/.gitignore [hunk 2]
```diff
@@ -0,0 +1,2 @@
+node_modules
```
Comment (mvdan) [inline controller/app/.gitignore:1]: might as well put this in the root gitignore

File: controller/controller.go [hunk 3]
```diff
@@ -1,3 +1,5 @@
+//go:generate go run github.com/filecoin-project/dealbot/controller/webutil/gen "./app" static/script.js
```
Comment (mvdan) [inline controller/controller.go:1]: a bit shorter:

> go run ./webutil/gen app static/script.js

File: controller/static/style.css [hunk 4]
```diff
@@ -0,0 +1,51 @@
...
+
+    .cron-picker-dow button.active {
+        background: #0c63e4;
+    }
+  
+  .cron-picker-time select {
+      width: 60px;
+      display: inline-block;
+      margin: 0 5px;
+  }
```
Comment (mvdan) [inline controller/static/style.css:50]: the indentation here seems to be off?

File: controller/webutil/gen/gen.go [hunk 5]
```diff
@@ -0,0 +1,34 @@
...
+	cmd.Dir = os.Args[1]
+	err := cmd.Run()
+	if err != nil {
+		fmt.Printf("Failed to install frontend dependencies: %v\n", err)
+		if _, err := os.Stat(path.Join(os.Args[1], "node_modules")); os.IsNotExist(err) {
+			os.Exit(1)
+		}
+	}
+
+	data := webutil.Compile(os.Args[1], true)
```
Comment (mvdan) [inline controller/webutil/gen/gen.go:29]: I wonder if it might be easier to just plop a `gen.go` file in the output directory that has:

```
//go:generate npm install
//go:generate go run github.com/evanw/esbuild/cmd/esbuild [whatever args]
```

This is assuming that what we're doing with esbuild could also be done via the command line. Since we're requiring npm to be installed anyway, I don't think calling it via os/exec gains us that much. I'll ultimately defer to you though, as I've never used esbuild.

File: controller/webutil/gen/gen.go [hunk 6]
```diff
@@ -0,0 +1,34 @@
...
+	if len(os.Args) < 3 {
+		fmt.Printf("%s\n", data)
+	}
+	ioutil.WriteFile(os.Args[2], []byte(data), 0644)
```
Comment (mvdan) [inline controller/webutil/gen/gen.go:33]: nit: error check

File: controller/app/index.js [hunk 7]
```diff
@@ -0,0 +1,93 @@
...
+    })
+
+    if (!$('#newRepeat').is(':checked')) {
+        $("#setschedule").hide();
+    }
+    $("#newRepeat").on('change', () => {
+        if ($('#newRepeat').is(':checked')) {
+            $("#setschedule").show();
+        } else {
+            $("#setschedule").hide();
```
Comment (mvdan) [inline controller/app/index.js:30]: isn't there a "toggle show/hide" method, or something like "set hidden as per this boolean"?

**Review by mvdan — COMMENTED** [review]
The generate/docker/Go side SGTM.

The JS/HTML part looks fine too, though my experience there is more limited.

File: Dockerfile (hunk 1 above)
Comment (mvdan) [reply Dockerfile:18]: I worded my first sentence weird. I mean that the output of `go generate` should always be in master, so I don't think re-running go generate gets us anywhere.

Speaking of which, I don't get the `go get -d -v ./...` either. Modules will download dependencies as needed in the next step already.

Comment (willscott) [reply Dockerfile:18]: I feel like `script.js` as a generated code that changes with every commit to frontend code is easier to build for docker rather than have in each commit as a way to not have to deal with merge conflicts

File: controller/webutil/gen/gen.go (hunk 5 above)
Comment (willscott) [reply controller/webutil/gen/gen.go:29]: This setup lets me run a development version of the controller that builds the frontend code from source every time it serves it, rather than using the pre-built version.

I can finish putting that in place so it's clear why i'm structuring it this way

File: controller/app/index.js (hunk 7 above)
Comment (willscott) [reply controller/app/index.js:30]: possibly?

File: Dockerfile (hunk 1 above)
Comment (mvdan) [reply Dockerfile:18]: Ah, I see. Maybe move it out of `go generate`, and into a build script of some sort, then? Because it is generally understood that the output of `go generate ./...` should be committed, so this will end up being confusing or problematic sooner or later. You can include its output dir in gitignore, though that adds to the confusion, if you ask me :)

File: controller/webutil/gen/gen.go (hunk 5 above)
Comment (mvdan) [reply controller/webutil/gen/gen.go:29]: Ah I see, makes sense.

File: Dockerfile (hunk 1 above)
Comment (willscott) [reply Dockerfile:18]: the output is in gitignore :)

File: controller/app/index.js [hunk 8]
```diff
@@ -0,0 +1,94 @@
...
+
+        if ($('#newRepeat').is(':checked')) {
+            data.Schedule = $('#newSchedule').val()
+            data.ScheduleLimit = $('#newScheduleLimit').val()
+        }
+        if ($('#newScheduleTag').val() !='') {
+            data.Tag = $('#newScheduleTag').val()
+        }
+
+        $.ajax({
```
Comment (hannahhoward) [inline controller/app/index.js:85]: I wonder: for $.ajax, maybe it isn't worth the jquery? $.ajax was just a thing that made things easier before `fetch` existed. Fetch returns a promise and then you can just do Promise.All on the an array of the promises and avoid the weird done/remaining log.

Non-blocking.

File: controller/state/statedb.go [hunk 9]
```diff
@@ -752,7 +752,7 @@ func (s *stateDB) PublishRecordsFrom(ctx context.Context, worker string) error {
 			if err != nil {
 				return err
 			}
-			tskBuilder := tasks.Type.FinishedTask.NewBuilder()
+			tskBuilder := tasks.Type.FinishedTask__Repr.NewBuilder()
```
Comment (hannahhoward) [inline controller/state/statedb.go:755]: This seems quite unrelated? What does it fix?

File: controller/controller.go [hunk 10]
```diff
@@ -121,7 +121,7 @@ func (fw *logEcapsulator) Write(p []byte) (n int, err error) {
 //go:embed static
 var static embed.FS
 
-func NewWithDependencies(listener, graphqlListener net.Listener, gqlToken string, recorder metrics.MetricsRecorder, backend state.State) (*Controller, error) {
+func NewWithDependencies(ctx *cli.Context, listener, graphqlListener net.Listener, recorder metrics.MetricsRecorder, backend state.State) (*Controller, error) {
```
Comment (hannahhoward) [inline controller/controller.go:124]: Nit: the purpose of seperating New from NewWithDependencies was to put the CLI parsing in a single place, so that tests wouldn't worry about assembling a CLI. Technically, NewWithDepedencies could be a private function and the tests could skip the _test underscore in the package name

**Review by hannahhoward — COMMENTED** [review]
Left comments -- mostly non-blocking but I would like to understand what the change to state.go is about, since that seems unrelated.

File: controller/state/statedb.go (hunk 9 above)
Comment (willscott) [reply controller/state/statedb.go:755]: I think it was either @mvdan or @rvagg  who mentioned that we should be de-serializing into __Repr's as well as serializing into them. This is a remnant of the constant 'the default schema'd types are not the ones you actually should be using problem in the current ipld-prime' 
in practice - it shouldn't matter any any valid representation _should_ de-serialize into the nonrepresentation type, but i think that might not be guaranteed?

**Review by hannahhoward — APPROVED** [review]

File: controller/state/statedb.go (hunk 9 above)
Comment (mvdan) [reply controller/state/statedb.go:755]: This was me, yes. In practice, in this particular case, it might not matter. But it's certainly asking for trouble. This would definitely break if we added a struct field with non-default representation, for example.
